### PR TITLE
[CARBONDATA-3064] Support separate audit log

### PR DESCRIPTION
**Pull request by jackylk**
A new audit log is implemented as following:
1. a framework is added for carbon command to record the audit log automatically, see command/package.scala
2. Audit logs are output by Auditor.java, log4j config example is provided in Auditor.java file comment
3. old audit log is removed

 - [X] Any interfaces changed?
 Internal interface change

 - [X] Any backward compatibility impacted?
 Yes

 - [X] Document update required?
Yes

 - [X] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [X] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1178/

File: processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+  public static void logOperationSuccess(String opName, String opId, String table,
+      String opTime, Map<String, String> extraInfo) {
+    OpEndMessage message = new OpEndMessage(opName, opId, table, opTime,
+        OpStatus.SUCCESS, extraInfo);
+    Gson gson = new GsonBuilder().disableHtmlEscaping().create();
+    String json = gson.toJson(message);
+    LOGGER.log(AuditLevel.AUDIT, json);
+  }
+
+  public static void logOperationFailed(String opName, String opId, String table,
```
Comment (xuchuanyin) [inline processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java]: find many duplicated code with `logOperationSuccess`

File: processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    try {
+      username = UserGroupInformation.getCurrentUser().getShortUserName();
+    } catch (IOException e) {
+      username = "unknown";
+    }
+  }
+
+  public static void logOperationStart(String opName, String opId) {
+    OpStartMessage message = new OpStartMessage(opName, opId);
+    Gson gson = new GsonBuilder().disableHtmlEscaping().create();
```
Comment (xuchuanyin) [inline processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java]: What if the message contains quota strings? Will we escape that?

File: processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    String json = gson.toJson(message);
+    LOGGER.log(AuditLevel.AUDIT, json);
+  }
+
+  public static void logOperationSuccess(String opName, String opId, String table,
+      String opTime, Map<String, String> extraInfo) {
+    OpEndMessage message = new OpEndMessage(opName, opId, table, opTime,
+        OpStatus.SUCCESS, extraInfo);
+    Gson gson = new GsonBuilder().disableHtmlEscaping().create();
```
Comment (Sssan520) [inline processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java]: maybe it is better to use the object gson as a member of the class Audit

File: processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+  // log message for operation start, it is written as a JSON record in the audit log
+  private static class OpStartMessage {
+    private String time;
+    private String username;
+    private String opName;
+    private String opId;
+    private OpStatus opStatus;
+
+    OpStartMessage(String opName, String opId) {
+      SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
```
Comment (Sssan520) [inline processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java]: maybe it is better to use the object format as a static member of the class OpStartMessage

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1391/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9442/

File: processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java (hunk 4 above)
Comment (jackylk) [reply processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java]: SimpleDateFormat is not thread safe, and this method maybe called in multiple thread

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/package.scala [hunk 5]
```diff
@@ -60,21 +63,71 @@ trait DataProcessOperation {
...
+  private var auditInfo: java.util.Map[String, String] = new java.util.HashMap[String, String]()
+
+  // holds the dbName and tableName for which this command is executed for
+  // used for audit log, set by subclass of AtomicRunnableCommand
+  private var table: String = _
+
+  // implement by subclass, return the operation name that record in audit log
+  protected def opName: String
+
+  protected def opTime(startTime: Long) = s"${(System.nanoTime() - startTime) / 1000L / 1000L} ms"
```
[the remark this answers is not in the document]
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/package.scala]: ok

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableAddHivePartitionCommand.scala [hunk 6]
```diff
@@ -55,6 +55,8 @@ case class CarbonAlterTableAddHivePartitionCommand(
 
   override def processMetadata(sparkSession: SparkSession): Seq[Row] = {
     table = CarbonEnv.getCarbonTable(tableName)(sparkSession)
+    setAuditTable(table)
+    setAuditInfo(Map("partition" -> partitionSpecsAndLocs.mkString(",")))
```
[the remark this answers is not in the document]
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableAddHivePartitionCommand.scala]: ok

File: processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java (hunk 1 above)
Comment (jackylk) [reply processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java]: ok, changed to logOperationEnd

File: processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java (hunk 2 above)
Comment (jackylk) [reply processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java]: I think gson will keep the quota string.
In gson document, it says "By default, Gson escapes HTML characters such as < > etc. Use this option to configure Gson to pass-through HTML characters as is."

File: processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java (hunk 3 above)
Comment (jackylk) [reply processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java]: ok, seems GSON is thread safe, so I change it to static member

File: processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java (hunk 4 above)
Comment (jackylk) [reply processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java]: ok, will fix

File: processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java (hunk 3 above)
Comment (Sssan520) [reply processing/src/main/java/org/apache/carbondata/processing/util/Auditor.java]: > ok, seems GSON is thread safe, so I change it to static member
yes there is a fix pr: https://github.com/google/gson/issues/63

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1433/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1218/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9482/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1220/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1435/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9484/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1471/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9520/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1262/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1265/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9529/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1480/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9570/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1312/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1523/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1530/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9577/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1319/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1534/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1323/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9582/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1324/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9583/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1535/

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1330/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9589/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1541/

**Comment by xuchuanyin** [comment]
LGTM

**Pull request closed without merging**